src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -1065,7 +1035,45 @@ void InterpreterCreateQuery::setEngine(ASTCreateQuery & create) const
             create.set(create.as_table_function, as_create.as_table_function->ptr());
         else
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot set engine, it's a bug.");
+    };
 
+    if (create.temporary)
+    {
+        if (!create.cluster.empty())
+            throw Exception(ErrorCodes::INCORRECT_QUERY, "Temporary tables cannot be created with ON CLUSTER clause");
+
+        if (create.as_table.empty())
```

Comment:
1) I think we should first check if engine was passed explicitly in `CREATE TEMPORARY ... ENGINE=...`, in that case it should have precedence to `AS table`
2) not sure if it's ok to overwrite engine=Memory with engine from as_table (it can lead to surprisizing effects to many users who are not even aware that temporary table can have different engines).


May be the sequence should be the following
1) if engine in passed explicitly - use it
2) if the default_temporary_table_engine like '%MergeTree%' then inherit the engine properties from as_table (not sure what't the bese option if default_temporary_table_engine is MergeTree and as_table is ReplacingMergeTree).
3) otherwise just use default_temporary_table_engine

@alexey-milovidov WDYT?